### ISSUE TITLE
Union of Literals as key in union of TypedDict false positive

### ISSUE THREAD

**Issue by connesy** [python/mypy#14459]
**Bug Report**
The function
```python
def get_value(key: CombinedDictKey, mapping: CombinedDict):
    return mapping[key]
```
raises an error when `mapping` is a union of two `TypedDict`s and `key` is a union of two `Literals` containing the keys of those `TypedDict`s. It works when typed with just one of the `Literal`s and `TypedDict`s.

**To Reproduce**

```python
from typing import Literal, TypedDict, Union

class DictA(TypedDict):
    x: int
    y: int

class DictB(TypedDict):
    z: str

class DictAB(DictA, DictB):
    pass

DictAKey = Literal["x", "y"]
DictBKey = Literal["z"]
DictABKey = Union[DictAKey, DictBKey]

dict_a = DictA(x=1, y=2)
dict_b = DictB(z="z")
dict_ab = DictAB(**dict_a, **dict_b)  # type: ignore  # Unrelated issue #11108

def get_value_from_A(key: DictAKey, mapping: DictA):
    return mapping[key]  # Passes type check as expected

def get_value_from_B(key: DictBKey, mapping: DictB):
    return mapping[key]  # Passes type check as expected

def get_value_from_AB(key: DictABKey, mapping: DictAB):
    return mapping[key]  # Fails type check with error: TypedDict key must be a string literal
```

**Expected Behavior**
`return mapping[key]` passes typing in all three cases.

**Actual Behavior**
In the case of `get_value_from_AB`, `return mapping[key]` raises `error: TypedDict key must be a string literal; expected one of ("z", "x", "y")`. This is unexpected as it passes for the functions typed with only one of `DictA` or `DictB`.

**Your Environment**

- Mypy version used: 0.991
- Python version used: 3.11.0

**Comment by asottile-sentry** [comment]
this fixes your example fwiw:

```diff
$ diff -u t.py{,.new}
--- t.py	2025-02-18 13:13:48
+++ t.py.new	2025-02-18 13:14:03
@@ -12,7 +12,7 @@
 
 DictAKey = Literal["x", "y"]
 DictBKey = Literal["z"]
-DictABKey = Union[DictAKey, DictBKey]
+DictABKey = Literal[DictAKey, DictBKey]
 
 dict_a = DictA(x=1, y=2)
 dict_b = DictB(z="z")
```

**Comment by asottile-sentry** [comment]
outside quick analysis if someone wants to fix the `Union[Lit1, Lit2]` is the same as `Literal[Lit1, Lit2]`

codepath appears to be here: https://github.com/python/mypy/blob/87704cd235cca2058f244d303ee5b5918e95c62a/mypy/checkexpr.py#L4642

a _guess_ is that maybe `get_proper_types(Union[Of, Literals])` should simplify somewhere earlier on in analysis (which I believe would allow this to pass)

**Comment by connesy** [comment]
> this fixes your example fwiw:

Thanks!

**Comment by dogtopus** [comment]
Not sure if it's related, but I'm getting something similar to this:

```python
from typing import Literal, reveal_type


type TypeA = dict[int, Literal['a', 'b', 'c']]
type TypeB = dict[Literal['key1', 'key2', 'key3'], float]
type TypeAB = TypeA | TypeB


a: TypeA = reveal_type({1: 'a'})  # OK.
b: TypeB = reveal_type({'key2': 12345.6})  # OK.
ab: TypeAB = reveal_type({1: 'a'})  # NG. Inferred type has no Literal[] anywhere.
ab2: TypeAB = reveal_type({'key2': 12345.6})  # NG. Inferred type has no Literal[] anywhere.
ab3: TypeAB = a  # OK.
```

Running mypy on it gives me

```
../test.py:9: note: Revealed type is "dict[int, Literal['a'] | Literal['b'] | Literal['c']]"
../test.py:10: note: Revealed type is "dict[Literal['key1'] | Literal['key2'] | Literal['key3'], float]"
../test.py:11: error: Incompatible types in assignment (expression has type "dict[int, str]", variable has type "dict[int, Literal['a', 'b', 'c']] | dict[Literal['key1', 'key2', 'key3'], float]")  [assignment]
../test.py:11: note: Revealed type is "dict[int, str]"
../test.py:12: error: Incompatible types in assignment (expression has type "dict[str, float]", variable has type "dict[int, Literal['a', 'b', 'c']] | dict[Literal['key1', 'key2', 'key3'], float]")  [assignment]
../test.py:12: note: Revealed type is "dict[str, float]"
Found 2 errors in 1 file (checked 1 source file)
```

The workaround suggested by @asottile-sentry obviously does not apply here either.

Pyright can correctly infer this type of typing.

mypy 1.20, Python 3.14.3
